### PR TITLE
Add reworked tests for Plan Adoption Leave

### PR DESCRIPTION
**Pull request by gclssvglx**
[Trello](https://trello.com/c/xit50RWt/629-smart-answer-flow-tests-plan-adoption-leave)

This converts the Plan Adoption Leave tests to use the node based testing
approach introduced in [ADR-3][1].

[1]: https://github.com/alphagov/smart-answers/blob/main/docs/arch/003-testing-flows.md

⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

### CODE REVIEW

File: test/flows/plan_adoption_leave_flow_test.rb [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+  context "question: child_match_date?" do
+    setup { testing_node :child_match_date? }
+
+    should "render the question" do
+      assert_rendered_question
+    end
+
+    context "next_node" do
+      should "have a next node of child_arrival_date? for a response of 3 months ago" do
```
Comment (kevindew) [inline test/flows/plan_adoption_leave_flow_test.rb:21]:
```suggestion
      should "have a next node of child_arrival_date?" do
```

Since it doesn't matter what the response is to reach that node

File: test/flows/plan_adoption_leave_flow_test.rb [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    should "render the question" do
+      assert_rendered_question
+    end
+
+    context "next_node" do
+      should "have a next node of leave_start? for a response that is after the child_match_date" do
+        assert_next_node :leave_start?, for_response: 1.day.after(3.months.ago).to_s
+      end
+
+      context "validation" do
```
Comment (kevindew) [inline test/flows/plan_adoption_leave_flow_test.rb:42]: this context is conventionally before `context "next_node"` as per: https://github.com/alphagov/smart-answers/blob/8d4d04d61c3ede76358fb10cf2470e94d4fa68a1/test/flows/state_pension_age_flow_test.rb#L35-L47

File: test/flows/plan_adoption_leave_flow_test.rb [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    should "render the question" do
+      assert_rendered_question
+    end
+
+    context "next_node" do
+      should "have a next node of adoption_leave_details for a response that is within 14 days of the child_arrival_date" do
+        assert_next_node :adoption_leave_details, for_response: 14.days.ago(3.months.since).to_s
+      end
+
+      context "validation" do
```
Comment (kevindew) [inline test/flows/plan_adoption_leave_flow_test.rb:70]: As per previous comment about this block placement.

File: test/flows/plan_adoption_leave_flow_test.rb [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+                    child_arrival_date?: 3.months.since.to_s
+    end
+
+    should "render the question" do
+      assert_rendered_question
+    end
+
+    context "next_node" do
+      should "have a next node of adoption_leave_details for a response that is within 14 days of the child_arrival_date" do
+        assert_next_node :adoption_leave_details, for_response: 14.days.ago(3.months.since).to_s
```
Comment (kevindew) [inline test/flows/plan_adoption_leave_flow_test.rb:67]: The `14.days.ago(3.months.since)` is quite tough to understand. I think it'd better with a separate variable a la:

```suggestion
        adoption_leave_start = 3.months.since - 14.days
        assert_next_node :adoption_leave_details, for_response: adoption_leave_start.to_s
```

File: test/flows/plan_adoption_leave_flow_test.rb [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+      end
+
+      context "validation" do
+        should "be invalid if the response is not within 14 days of the child_arrival_date" do
+          assert_invalid_response 15.days.ago(3.months.since).to_s
```
Comment (kevindew) [inline test/flows/plan_adoption_leave_flow_test.rb:72]: And as per previous comment on this logic being quite hard to parse.

**Review by kevindew — COMMENTED** [review]
This looks good too, similar nitpicky comments.

**Comment by gclssvglx** [comment]
Thanks for the comment and suggestions @kevindew 👍  All fixed up

**Review by kevindew — APPROVED** [review]
